lib/AppHelper.php

```diff
@@ -448,4 +450,23 @@ public static function formatDateTime($datetime, $datetype, $timetype, $locale=n
 		return $dateFormatter->format($datetime);
 	}
 
+	/**
+	 * Waits for a test to satisfy (i.e. to return a truthy value)
+	 *
+	 * @param Closure $test -- test to run. Must return truthy on success.
+	 * @param int $timeout -- seconds to giving up waiting
+	 * @param int $interval -- seconds between retry attempts
+	 * @return bool -- true if test succeeded within timeout, false otherwise
+	 */
+	public static function waitFor($test, $timeout=60, $interval=1) {
+		$startTime = time();
+		do {
+			$lastTry = time();
+			if($test()) return true;
+			while(time() < $lastTry+$interval) sleep(1);
```

Comment:
1. I think the code is more correct without `sleep()`, because we are actively waiting until expected timeout takes place. However, busy-waiting or busy-looping is not very interesting as it requires CPU resources, but it is surely acceptable for experimental features for example, which is the case here given your comment above.
2. From my understanding, `sleep()` pauses the main thread (the whole program), and this is probably not what we want here...

Reply:
If you're fine with using `sleep()` in the `waitFor()` function, which seems to be the case given your last comment, then the `while` loop inside the `do-while` loop should be removed. Here is an alternative for `waitFor()` followed by a simple test code.

```php
function waitFor($test, $timeout=60, $interval=1) {
    $startTime = time();
    $minWaitDuration = min($timeout, $interval);
    do {
        if($test()) return true;
        sleep($minWaitDuration);
    }
    while(time() < $startTime+$timeout);
    return false;
}

// simple use case
$count = 0;
waitFor(function() use (&$count) {
    $finished = (++$count === 3);
    $msg = "Call $count: " . ($finished ? 'done!' : 'in progress...');
    echo $msg, PHP_EOL;
    return $finished;
});
echo 'Terminated!', PHP_EOL;
```

- You can also pass 1 as `$timeout` and 2 as `$interval` to `waitFor()` in the above example. You will see that `waitFor()` terminates immediately when `$timeout` runs out before `$interval`; so the rephrasing suggested earlier becomes obsolete (see below).
    - _Note: `$interval` timeout is checked before `$timeout` timeout, so the latter will not be detected until the first one takes place (e.g. when `$interval` is greater than `$timeout`)_
- Moreover, if you do not want to check `$test()` when `$timeout` is 0, then the `do-while` loop wan be converted to a `while` loop.
- Finally, the test code can be provided in the documentation of the function as a simple use case.